File: python/combination_sum_ii.py
class Solution:
    def combinationSum2(self, candidates: list[int], target: int) -> list[list[int]]:
        """time complexity: O(n * 2^n)"""
        subset: list[int] = []
        result: list[list[int]] = []
        candidates.sort()
        self.dfs(0, 0, subset, result, candidates, target)
        return result

    def dfs(
        self,
        i: int,
        accu: int,
        subset: list[int],
        result: list[list[int]],
        candidates: list[int],
        target: int,
    ) -> None:
        if accu == target:
            result.append(subset.copy())
            return
        if i >= len(candidates) or accu > target:
            return
        subset.append(candidates[i])
        self.dfs(i + 1, accu + candidates[i], subset, result, candidates, target)
        subset.pop()
        while i < len(candidates) - 1 and candidates[i] == candidates[i + 1]:
            i += 1
        self.dfs(i + 1, accu, subset, result, candidates, target)


def test_combinationSum2_case_1():
    # arrange
    candidates: list[int] = [10, 1, 2, 7, 6, 1, 5]
    target: int = 8
    expected: list[list[int]] = [[1, 1, 6], [1, 2, 5], [1, 7], [2, 6]]

    # act
    solution = Solution()
    actual = solution.combinationSum2(candidates, target)

    # assert
    expected.sort()
    actual.sort()
    assert expected == actual


def test_combinationSum2_case_2():
    # arrange
    candidates: list[int] = [2, 5, 2, 1, 2]
    target: int = 5
    expected: list[list[int]] = [[1, 2, 2], [5]]

    # act
    solution = Solution()
    actual = solution.combinationSum2(candidates, target)

    # assert
    expected.sort()
    actual.sort()
    assert expected == actual

File: rust/src/code/combination_sum_ii.rs
pub struct Solution {}

impl Solution {
    pub fn combination_sum2(mut candidates: Vec<i32>, target: i32) -> Vec<Vec<i32>> {
        candidates.sort();
        let mut result: Vec<Vec<i32>> = Vec::new();
        let mut subset: Vec<i32> = Vec::new();
        Self::dfs(0, &candidates, &mut subset, &mut result, target);

        return result;
    }

    fn dfs(
        pos: usize,
        candidates: &Vec<i32>,
        subset: &mut Vec<i32>,
        result: &mut Vec<Vec<i32>>,
        target: i32,
    ) {
        if target == 0 {
            result.push(subset.clone());
        } else if target < 0 {
            return;
        }

        let mut prev: i32 = -1;
        for i in pos..candidates.len() {
            if candidates[i] == prev {
                continue;
            }
            subset.push(candidates[i]);
            Self::dfs(i + 1, candidates, subset, result, target - candidates[i]);
            subset.pop().unwrap();
            prev = candidates[i];
        }
    }
}

#[cfg(test)]
mod tests {
    use super::Solution;

    #[test]
    fn combination_sum2_case_1() {
        // arrange
        let candidates: Vec<i32> = vec![10, 1, 2, 7, 6, 1, 5];
        let target: i32 = 8;
        let mut expected: Vec<Vec<i32>> =
            vec![vec![1, 1, 6], vec![1, 2, 5], vec![1, 7], vec![2, 6]];

        // act
        let mut actual = Solution::combination_sum2(candidates, target);

        // assert
        expected.sort();
        actual.sort();
        assert_eq!(expected, actual);
    }

    #[test]
    fn combination_sum2_case_2() {
        // arrange
        let candidates: Vec<i32> = vec![2, 5, 2, 1, 2];
        let target: i32 = 5;
        let mut expected: Vec<Vec<i32>> = vec![vec![1, 2, 2], vec![5]];

        // act
        let mut actual = Solution::combination_sum2(candidates, target);

        // assert
        expected.sort();
        actual.sort();
        assert_eq!(expected, actual);
    }
}
